pennylane/math/matrix_manipulation.py

```diff
@@ -102,105 +103,134 @@ def expand_matrix(base_matrix, wires, wire_order=None, sparse_format="csr"):
+        else:
+            mat = kron_interface(mat, eye_interface(len(wire_difference)))
+
+    # permute matrix
+    if interface == "scipy":
+        mat = _permute_sparse_matrix(mat, expanded_wires, subset_wire_order)
+    else:
+        mat = _permute_dense_matrix(mat, expanded_wires, subset_wire_order, batch_dim)
 
+    # expand the matrix even further if needed
+    if len(expanded_wires) < len(wire_order):
+        mats = []
+        num_pre_identities = min(wire_indices)
+        if num_pre_identities > 0:
+            mats.append((eye_interface(num_pre_identities),))
 
-def _sparse_expand_matrix(base_matrix, wires, wire_order, format="csr"):
-    """Re-express a sparse base matrix acting on a subspace defined by a set of wire labels
-    according to a global wire order.
+        mats.append(tuple(mat) if batch_dim else (mat,))
+
+        num_post_identities = len(wire_order) - max(wire_indices) - 1
+        if num_post_identities > 0:
+            mats.append((eye_interface(num_post_identities),))
+
+        # itertools.product will create a tuple of matrices for each different batch
+        mats_list = list(itertools.product(*mats))
+        # here we compute the kron product of each different tuple and stack them back together
+        expanded_batch_matrices = [reduce(kron_interface, mats) for mats in mats_list]
+        mat = (
+            qml.math.stack(expanded_batch_matrices, like=interface)
+            if len(expanded_batch_matrices) > 1
+            else expanded_batch_matrices[0]
+        )
+    return mat.asformat(sparse_format) if interface == "scipy" else mat
+
+
+def _permute_sparse_matrix(matrix, wires, wire_order):
+    """Permute the matrix to match the wires given in `wire_order`.
 
     Args:
-        base_matrix (scipy.sparse.spmatrix): base matrix to expand
-        wires (Iterable): wires determining the subspace that base matrix acts on; a base matrix of
+        matrix (scipy.sparse.spmatrix): matrix to permute
+        wires (list): wires determining the subspace that base matrix acts on; a base matrix of
             dimension :math:`2^n` acts on a subspace of :math:`n` wires
-        wire_order (Iterable): global wire order, which has to contain all wire labels in ``wires``, but can also
-            contain additional labels
-        format (str): string representing the preferred scipy sparse matrix format to cast the expanded
-            matrix too
+        wire_order (list): global wire order, which has to contain all wire labels in ``wires``,
+            but can also contain additional labels
 
     Returns:
-        tensor_like: expanded matrix
+        scipy.sparse.spmatrix: permuted matrix
     """
-    base_matrix.eliminate_zeros()
-    wires = wires.tolist() if isinstance(wires, qml.wires.Wires) else list(copy.copy(wires))
+    U = _permutation_sparse_matrix(wires, wire_order)
+    if U is not None:
+        matrix = U.T @ matrix @ U
+        matrix.eliminate_zeros()
```

Comment:
Why do we need to use eliminate zeros?

Reply:
I realised that in some cases there were a lot of sparse coefficients that were equal to 0... I really don't know where they come from, that's why I added the `eliminate_zeros` method after every operation.